akka-http-core/src/main/scala/akka/http/engine/ws/Websocket.scala

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (C) 2009-2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+
+package akka.http.engine.ws
+
+import akka.http.model.ws.{ BinaryMessage, TextMessage, Message }
+import akka.stream.scaladsl._
+import akka.stream.{ FlowShape, FanOutShape2 }
+import akka.http.util._
+import akka.http.model.ws._
+
+private[akka] object Websocket {
+  import FrameHandler._
+
+  def handleMessages[T](messageHandler: Flow[Message, Message, T], serverSide: Boolean = true): Flow[FramePart, FramePart, Unit] = {
+    val collectMessage: Flow[Source[MessageDataPart, Unit], Message, Unit] =
+      Flow[Source[MessageDataPart, Unit]]
+        .headAndTail
+        .map {
+          case (TextMessagePart(text, true), remaining) ⇒
+            // TODO: require(remaining.isEmpty)
+            TextMessage.Strict(text)
+          case (TextMessagePart(text, false), remaining) ⇒
+            TextMessage.Streamed(remaining.collect {
+              case t: TextMessagePart if t.data.nonEmpty ⇒ t.data
+            })
+          case (BinaryMessagePart(data, true), remaining) ⇒
+            // TODO: require(remaining.isEmpty)
+            BinaryMessage.Strict(data)
+          case (first @ BinaryMessagePart(data, false), remaining) ⇒
+            val parts = (Source.single(first) ++ remaining).mapMaterialized(_ ⇒ ())
+
+            BinaryMessage.Streamed(parts.collect {
+              case t: BinaryMessagePart if t.data.nonEmpty ⇒ t.data
+            })
+        }
+
+    val userFlow =
+      Flow[MessagePart]
+        .splitWhen(_.isMessageEnd) // FIXME using splitAfter from #16885 would simplify protocl a lot
+        .map(_.collect {
+          case m: MessageDataPart ⇒ m
+        })
+        .via(collectMessage)
+        .via(messageHandler)
+        .via(MessageToFrameRenderer.create(serverSide))
+
+    Flow[FramePart]
+      .via(printEvent("parserOut"))
+      .via(Masking.unmaskIf(serverSide))
+      .via(FrameHandler.create(server = serverSide))
+      .via(printEvent("handlerOut"))
+      .viaMat(StreamUtils.routeEither(Flow[FramePart], userFlow))(Keep.right)
+      .via(Masking.maskIf(!serverSide))
+      .via(printEvent("maskOut"))
```

Comment:
`printEvent` will be possible to replace with the build-in `log(name)` after https://github.com/akka/akka/pull/17162 is merged

// ticket for this task: https://github.com/akka/akka/issues/17184